### DISCUSSION TITLE
Performance optimisation for large numbers of routing calls from go.

### DISCUSSION THREAD

**Discussion by DigNZ** [valhalla/valhalla#3497] [Q&A] [answered]
To pre-seed data in my application I need to make hundreds of thousands of routing and elevation calls and populate a database with the results. I essentially need a matrix call but there are several hundred locations, I need locations to all other locations and I need the shape of each path as well.

The application is written in go, and I'm using the http interface to make these calls (to valhalla running local to the application). It works but its at the limit of what it can handle and I need it to be able to scale without taking days to run. The HTTP requests cause a bottleneck both in how long it takes to get a result and how many requests and be served per second. 

These requests can be run concurrently and gorountines provide a small boost.

Would using the C++ API directly be faster? Is there a way to do that using go?

**Comment by kevinkreiser** [comment, marked as answer]
We currently don't support any go bindings but I'm certain a c-wrapper could be written to allow for this via cgo. The real question is: *should* you do this, will it actually increase your throughput?

before we answer that question though let me mention that if you have the option of running valhalla *not* local to the application (ie on different machines) then what i would do, and what the most typical thing to do, is to spin up another valhalla http server on another machine and load balance requests between multiple http valhalla services via round robin or via a more intelligent proxy. in this way you can always just add more capacity when you have more demand or what to finish your work load faster. i dont know enough about  your application to be able to say if this is an option but this is by far the most typical way to achieve what you are looking to do.

anyway, if you cant do the above and you still want to know if directly calling the library is faster than http then i would suggest you do some tests to answer the question of whether its fast enough to be worth your while. Broadly what I would suggest is to run your application, capturing the requests you need to run to a file. Capture a significant sized set so we can be really sure that it accurately represents your use-case. From there I would hack together a simple program that has a threadpool, reads your requests files in and then lets the threads burn down the list. Each thread would have a local instance of an `actor_t` which you could use to answer the requests. If that simple program can burn down your request set, significantly faster than you can burn it down over http then it would be worth your while to figure out how to by pass http.

**Reply by DigNZ** [reply]
Thanks, that is a great idea - I'll set up a few valhalla servers and loadbalance and test that in the first instance. 

Happy to build a c-wrapper if it is going to make a significant difference but we are not yet sure that it will then multiple services sounds like the way to do.

**Reply by nilsnolde** [reply]
just a small addition, might be redundant, I'm pretty biased there: there's also python bindings/wrappers (ON by default), which I'd find easier to test what kevin described

**Reply by kevinkreiser** [reply]
true, if you are less comfortable with c++ this would be a good option. for completeness sake, such an implementation must use multiprocessing and must use the process initializer to create a long lived actor instance for each process. that would get the job done if you did want go the route of skipping http. i still think the best thing is to stay http and add more servers though :smile:

**Comment by purew** [comment]
Is the assumption here that the json serialization and deserialization takes a significant amount of time? That's the only part that potentially would be removed by having your go-code call the c++ code directly, right?

**Reply by kevinkreiser** [reply]
there is some overhead in the passing of state from the http server to the worker and from each worker to the next worker, as they do this over sockets and they do serialize and deserialize pbf each time. in multiprocess mode there is even more overhead (though also resiliancy) as running these things at the process level takes more cpu resources away from computing routes as well.

what i cant do off the top of my head though is quantify any of that :smile: but it does exist, in addition to the json serialization as you rightly mentioned.

**Comment by DigNZ** [comment]
My assumption - also unquantified - is that the json serialization and the http layer itself would add some overhead, but now that I think about it it is unlikely to be orders of magnitude. Let's say that overhead doubles the response time, then running several instances of valhalla and load balancing between them would probably be the better option anyway and is certainly simpler.  When I'm at a loose end I might try implementing the preprocessing in python or even C++ as a standalone tool (the results get stored into SQLite so it should be fine to process in a language other than Go) and see what effect that has.

I've tried three instances of valhalla and it's noticably improved the throughput. Every instance of valhalla is sharing the same data directory which may or may not be a good idea. 

I'll take a look at the application itself and see what I can do to reduce the amount of preprocessing that is actually needed in the first place. After that pre-calculation stage the application still makes a couple of calls to Valhalla to serve a request but its super quick by then.

**Reply by purew** [reply]
I'd also assume the routing calculations would account for the majority of compute time. Maybe double-check that you're compiling in release mode with optimizations enabled and that you run in memory-map mode and have enough ram?

**Reply by nilsnolde** [reply]
One last optimization would be to implement PBF in/out for matrix as well. It’s big JSON payloads even though surely dwarfed by the matrix algo unfortunately.

**Comment by DigNZ** [comment]
@purew I have compiled in release mode - how do I run in memory-map mode and do I need enough ram to completely contain the valhalla_tiles.tar or more than that?

**Reply by kevinkreiser** [reply]
@DigNZ the readme shows how to do this in its example.  you can turn your tile directory into a tar file and then set that tar file as the value to the `tile_extract` key in the config.

**Reply by purew** [reply]
> do I need enough ram to completely contain the valhalla_tiles.tar or more than that?

For good performance, you need enough ram to load the parts of the planet file that you query into memory. The OS will transparently load the parts of the file from disk into memory and remove unused parts when memory is filling up.

If you query the entiry planet of 60Gb, then you'll want > 60Gb of memory for good performance.

If you query only 5Gb of the 60Gb, you only need >5Gb for good performance.

**Reply by DigNZ** [reply]
Excellent thanks for that. So I have only built Oceania and Britain and Ireland at the moment as I only need coverage on those (and fortunately they are all islands which makes things nice and easy). This produced a ~3.5GB valhalla_tiles.tar file but >100GB of files for elevation data. I don't need elevation information on every request but I do on a number of requests. I'm thinking of running multiple nodes with shared access to the valhalla_tiles.tar and elevation data, load balance between them and think I could get away with 8GB of RAM per node with a larger instance for building tiles and keeping the data up to date...

**Reply by kevinkreiser** [reply]
sounds reasonable to me! if you feel like we've gotten to the bottom of your problem would you mind marking one of the posts as an answer so the discussion shows as answered? thanks again!

**Reply by DigNZ** [reply]
Thanks for all your help everyone.
